packages/e2e-tests/src/Yarn.ts

```diff
@@ -0,0 +1,128 @@
+import * as os from 'os';
+import * as path from 'path';
+import { Cache } from './Cache';
+import * as core from '@actions/core';
+import * as glob from '@actions/glob';
+import * as artifact from '@actions/artifact';
+import { SpawnSync } from './SpawnSync';
+import { Repository } from './Repository';
+
+class Yarn {
+	constructor(
+		private readonly repo: Repository,
+		private readonly spawnSync: SpawnSync,
+		private readonly cache: Cache
+	) {}
+
+	public async install({ frozenLockfile }: { frozenLockfile: boolean }): Promise<Yarn> {
+		const args = ['install'];
+
+		if (frozenLockfile) {
+			args.push('--frozen-lockfile');
+		}
+
+		await this.call(args, ['node_modules', '*/node_modules']);
+
+		return this;
+	}
+
+	public async build(): Promise<Yarn> {
+		await this.call(['build'], ['build']);
+		return this;
+	}
+
+	public async test(env: Record<string, string>): Promise<Yarn> {
+		// TODO: once e2e-tests package is extracted, update this
+
+		const caRoot = this.spawnSync.call('mkcert', ['-CAROOT'], { stdout: 'pipe' }).stdout.trim();
+
+		const reportPath = path.resolve(os.tmpdir(), 'playwright-report');
+
+		env['NODE_EXTRA_CA_CERTS'] = `${caRoot}/rootCA.pem`;
+
+		env['PLAYWRIGHT_HTML_REPORT'] = reportPath;
+
+		// if docker cache will become available, restore should be called here
+
+		const buffer = this.spawnSync.call('yarn', ['workspace', '@eventespresso/e2e', 'playwright', 'test'], {
+			env,
+		});
+
+		// if docker cache will become available, save should be called here
+
+		if (buffer.status !== 0) {
+			await this.saveTestReport(reportPath);
+		}
+
+		return this;
+	}
+
+	private async saveTestReport(reportPath: string): Promise<void> {
+		core.notice(`Saving Playwright report found at: ${reportPath}`);
+
+		const client = artifact.create();
+
+		const pattern = path.resolve(reportPath, '**/*');
+
+		const globber = await glob.create(pattern);
+
+		const files = await globber.glob();
+
+		const reportName = `playwright-report-run-${process.env.GITHUB_RUN_NUMBER}-attempt-${process.env.GITHUB_RUN_ATTEMPT}`;
+
+		let response = undefined;
+
+		try {
+			response = await client.uploadArtifact(reportName, files, reportPath, {
+				continueOnError: true,
```

Comment:
Even if Playwright report is not fully saved, we still have a meaningful outcome i.e. something fails so no need to halt the rest of the workflow